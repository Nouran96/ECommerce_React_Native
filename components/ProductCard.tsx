import { View, Image, StyleSheet, TouchableOpacity } from "react-native";
import { AntDesign } from "@expo/vector-icons";
import { Text } from "./Themed";
import Colors from "../constants/Colors";
import useColorScheme from "../hooks/useColorScheme";
import ColorView from "./ColorView";
import MainButton from "./MainButton";
import { useAppDispatch, useAppSelector } from "../types";
import {
  addProductToCart,
  decrementQuantity,
  removeProductFromCart,
} from "../store/cart/cartSlice";
import PriceTag from "./PriceTag";
import { useNavigation } from "@react-navigation/native";
import { toggleProductFromWishlist } from "../store/wishlist/wishlistSlice";

interface ProductCardProps {
  product: any;
  showCartControls?: boolean;
  showQuantity?: boolean;
  imageStyles?: {};
  showOnSaleTag?: boolean;
  showWishlistBtn?: boolean;
}

export default function ProductCard({
  product,
  showCartControls,
  showQuantity,
  imageStyles,
  showOnSaleTag,
  showWishlistBtn,
}: ProductCardProps) {
  const colorScheme = useColorScheme();
  const dispatch = useAppDispatch();
  const navigation = useNavigation();

  const {
    wishlist: { products },
  } = useAppSelector((state) => state);

  return (
    <TouchableOpacity
      onPress={() => navigation.navigate("ProductDetails", { product })}
    >
      <View style={styles.product}>
        <View style={{ position: "relative" }}>
          {product.images && product.images.length > 0 && (
            <View>
              <Image
                style={{ ...styles.productImage, ...imageStyles }}
                source={{ uri: product.images[0].url }}
              />
            </View>
          )}

          {showWishlistBtn && (
            <TouchableOpacity
              onPress={() => dispatch(toggleProductFromWishlist(product))}
              style={{ position: "absolute", right: 0, top: 0 }}
            >
              <AntDesign
                name={products[product.code] ? "heart" : "hearto"}
                size={30}
                color={Colors[colorScheme].tint}
              />
            </TouchableOpacity>
          )}

          {showOnSaleTag && product.sale && (
            <View
              style={{
                position: "absolute",
                left: 0,
                backgroundColor: Colors[colorScheme].tint,
                borderRadius: 5,
                paddingHorizontal: 5,
                paddingVertical: 2,
              }}
            >
              <Text style={{ color: "white" }}>on Sale</Text>
            </View>
          )}
        </View>

        <View style={styles.productContent}>
          <Text style={styles.title}>
            {product.name} {showQuantity && `x${product.quantity}`}
          </Text>
          <View style={{ flexDirection: "row" }}>
            {product.sale && (
              <PriceTag
                price={product.whitePrice.value * product.quantity}
                sale
              />
            )}

            <PriceTag
              price={
                (product.sale
                  ? product.redPrice.value
                  : product.whitePrice.value) * product.quantity
              }
            />
          </View>
          <View
            style={{
              flexDirection: "row",
              justifyContent: "space-between",
              width: "100%",
            }}
          >
            <Text style={styles.row}>Size: {product.selectedSize}</Text>
            <View style={styles.row}>
              <Text>Color: </Text>
              <ColorView
                color={product.selectedColor}
                otherStyles={{ width: 20, height: 20 }}
              />
            </View>
          </View>

          {showCartControls && (
            <View
              style={{
                flexDirection: "row",
                justifyContent: "space-between",
                alignItems: "center",
                width: "100%",
              }}
            >
              <View
                style={{
                  flexDirection: "row",
                  justifyContent: "space-between",
                  alignItems: "center",
                }}
              >
                <MainButton
                  title="-"
                  disabled={product.quantity === 1}
                  onPress={() => dispatch(decrementQuantity(product))}
                />

                <Text style={{ paddingHorizontal: 15 }}>
                  {product.quantity}
                </Text>
                <MainButton
                  title="+"
                  onPress={() => dispatch(addProductToCart(product))}
                />
              </View>

              <TouchableOpacity
                onPress={() => dispatch(removeProductFromCart(product))}
              >
                <AntDesign
                  name="delete"
                  size={24}
                  color={Colors[colorScheme].tint}
                />
              </TouchableOpacity>
            </View>
          )}
        </View>
      </View>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  product: {
    width: "100%",
    flexDirection: "row",
    alignItems: "center",
    marginVertical: 15,
  },
  productImage: {
    width: 150,
    height: 150,
    borderRadius: 30,
  },
  productContent: {
    alignItems: "flex-start",
    flexGrow: 1,
    flex: 1,
    marginHorizontal: 15,
  },
  title: {
    marginVertical: 10,
  },
  row: {
    flexDirection: "row",
    justifyContent: "flex-start",
    marginVertical: 5,
  },
});
