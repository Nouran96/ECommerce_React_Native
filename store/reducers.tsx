import sharedReducer from "./shared/sharedSlice";

const rootReducers = {
  shared: sharedReducer,
};

export default rootReducers;
