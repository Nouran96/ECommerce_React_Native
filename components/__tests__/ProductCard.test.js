import React from "react";
import { Provider } from "react-redux";
import { fireEvent, render, waitFor } from "@testing-library/react-native";

import ProductCard from "../ProductCard";
import store from "../../store";

const productMock = {
  code: "1024613_group_002",
  name: "Oversized Hooded Jacket",
  quantity: 1,
  stock: {
    stockLevel: 1,
  },
  price: {
    currencyIso: "USD",
    value: 29.99,
    priceType: "BUY",
    formattedValue: "$ 29.99",
    type: "WHITE",
  },
  images: [
    {
      url: "https://lp2.hm.com/hmgoepprod?set=source[/83/b3/83b356d257923dc276fc2138e29370e7bf22b7c1.jpg],origin[dam],category[],type[DESCRIPTIVESTILLLIFE],res[m],hmver[2]&call=url[file:/product/style]",
    },
  ],
  categories: [],
  pk: "9335892869121",
  whitePrice: {
    currencyIso: "USD",
    value: 29.99,
    priceType: "BUY",
    formattedValue: "$ 29.99",
    type: "WHITE",
  },
  articles: [
    {
      code: "1024613002",
      name: "Oversized Hooded Jacket",
      images: [
        {
          url: "https://lp2.hm.com/hmgoepprod?set=source[/83/b3/83b356d257923dc276fc2138e29370e7bf22b7c1.jpg],origin[dam],category[],type[DESCRIPTIVESTILLLIFE],res[m],hmver[2]&call=url[file:/product/style]",
        },
      ],
      pk: "9320789278721",
      whitePrice: {
        currencyIso: "USD",
        value: 29.99,
        priceType: "BUY",
        formattedValue: "$ 29.99",
        type: "WHITE",
      },
      logoPicture: [
        {
          url: "https://lp2.hm.com/hmgoepprod?set=source[/5b/60/5b606c83bd1d914f6d339db831c7f60481b8fdc2.jpg],origin[dam],category[],type[LOOKBOOK],res[m],hmver[1]&call=url[file:/product/main]",
        },
      ],
      normalPicture: [
        {
          url: "https://lp2.hm.com/hmgoepprod?set=source[/5b/60/5b606c83bd1d914f6d339db831c7f60481b8fdc2.jpg],origin[dam],category[],type[LOOKBOOK],res[m],hmver[1]&call=url[file:/product/main]",
        },
      ],
      markers: [
        {
          text: "Conscious choice  ",
          type: "ENVIRONMENT",
        },
      ],
      visible: false,
      numbersOfPieces: 0,
      ticket:
        "Oy9wbHAvcHJvZHVjdC1saXN0LXdpdGgtY291bnQvcHJvZHVjdC1saXN0OyM7cHJvZHVjdF9rZXk7MTAyNDYxM19ncm91cF8wMDJfZW5fdXM7MTAyNDYxMzAwMl9lbl91cztPQkpFQ1RJVkUkO05PTkU6Tk9ORTs0MDs",
      dummy: false,
      ecoTaxValue: 0,
      redirectToPdp: false,
      comingSoon: false,
      color: {
        code: "000000",
        text: "Black",
        filterName: "Black_000000",
        hybrisCode: "09",
      },
      rgbColor: "#272628",
      genArticle: "024613002225",
      environmentalMarkers: ["Conscious choice  "],
    },
  ],
  markers: [
    {
      text: "Conscious choice  ",
      type: "ENVIRONMENT",
    },
  ],
  visible: false,
  concept: ["BASICS ,DIVIDED"],
  numbersOfPieces: 0,
  defaultArticle: {
    code: "1024613002",
    name: "Oversized Hooded Jacket",
    images: [
      {
        url: "https://lp2.hm.com/hmgoepprod?set=source[/83/b3/83b356d257923dc276fc2138e29370e7bf22b7c1.jpg],origin[dam],category[],type[DESCRIPTIVESTILLLIFE],res[m],hmver[2]&call=url[file:/product/style]",
      },
    ],
    pk: "9320789278721",
    whitePrice: {
      currencyIso: "USD",
      value: 29.99,
      priceType: "BUY",
      formattedValue: "$ 29.99",
      type: "WHITE",
    },
    logoPicture: [
      {
        url: "https://lp2.hm.com/hmgoepprod?set=source[/5b/60/5b606c83bd1d914f6d339db831c7f60481b8fdc2.jpg],origin[dam],category[],type[LOOKBOOK],res[m],hmver[1]&call=url[file:/product/main]",
      },
    ],
    normalPicture: [
      {
        url: "https://lp2.hm.com/hmgoepprod?set=source[/5b/60/5b606c83bd1d914f6d339db831c7f60481b8fdc2.jpg],origin[dam],category[],type[LOOKBOOK],res[m],hmver[1]&call=url[file:/product/main]",
      },
    ],
    markers: [
      {
        text: "Conscious choice  ",
        type: "ENVIRONMENT",
      },
    ],
    visible: false,
    numbersOfPieces: 0,
    ticket:
      "Oy9wbHAvcHJvZHVjdC1saXN0LXdpdGgtY291bnQvcHJvZHVjdC1saXN0OyM7cHJvZHVjdF9rZXk7MTAyNDYxM19ncm91cF8wMDJfZW5fdXM7MTAyNDYxMzAwMl9lbl91cztPQkpFQ1RJVkUkO05PTkU6Tk9ORTs0MDs",
    dummy: false,
    ecoTaxValue: 0,
    redirectToPdp: false,
    comingSoon: false,
    color: {
      code: "000000",
      text: "Black",
      filterName: "Black_000000",
      hybrisCode: "09",
    },
    rgbColor: "#272628",
    genArticle: "024613002225",
    environmentalMarkers: ["Conscious choice  "],
  },
  sale: false,
  variantSizes: [
    {
      orderFilter: 369,
      filterCode: "M",
    },
    {
      orderFilter: 366,
      filterCode: "S",
    },
    {
      orderFilter: 357,
      filterCode: "XXS",
    },
    {
      orderFilter: 372,
      filterCode: "XL",
    },
    {
      orderFilter: 363,
      filterCode: "XS",
    },
    {
      orderFilter: 375,
      filterCode: "XXL",
    },
    {
      orderFilter: 370,
      filterCode: "L",
    },
  ],
  swatches: [],
  articleCodes: ["1024613002", "1024613001", "1024613006", "1024613007"],
  ticket:
    "Oy9wbHAvcHJvZHVjdC1saXN0LXdpdGgtY291bnQvcHJvZHVjdC1saXN0OyM7cHJvZHVjdF9rZXk7MTAyNDYxM19ncm91cF8wMDJfZW5fdXM7MTAyNDYxMzAwMl9lbl91cztPQkpFQ1RJVkUkO05PTkU6Tk9ORTs0MDs",
  searchEngineProductId: "1024613_group_002_en_us",
  dummy: false,
  linkPdp: "/en_us/productpage.1024613002.html",
  categoryName: "Ladies",
  rgbColors: ["#272628", "#EEEDE1", "#3C2A2B", "#515359"],
  articleColorNames: ["Black", "White", "Dark brown", "Dark gray"],
  ecoTaxValue: 0,
  swatchesTotal: 5,
  showPriceMarker: false,
  redirectToPdp: false,
  mainCategoryCode: "ladies_basics_tops_longsleeve",
  comingSoon: false,
  brandName: "H&M",
};

const ProductCardWithWrapper = (props) => {
  return (
    <Provider store={store}>
      <ProductCard {...props} />
    </Provider>
  );
};

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
    }),
  };
});

// jest.mock("../../store/cart/cartSlice.tsx", () => ({
//   addProductToCart: jest.fn(),
// }));

describe("ProductCard", () => {
  it("should render correctly", () => {
    const { toJSON } = render(<ProductCardWithWrapper product={productMock} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("should render product name correctly", () => {
    const { getByText } = render(
      <ProductCardWithWrapper product={productMock} />
    );

    expect(getByText(productMock.name)).toBeDefined();
  });

  it("should increase quantity when button clicked", () => {
    const { getByText, findByTestId } = render(
      <ProductCardWithWrapper product={productMock} showCartControls />
    );

    const button = getByText("+");

    fireEvent.press(button);

    waitFor(() => {
      const quantityEl = findByTestId("quantity");
      expect(store.getState().cart.products[productMock.code].quantity).toBe(
        productMock.quantity + 1
      );
    });
  });
});
