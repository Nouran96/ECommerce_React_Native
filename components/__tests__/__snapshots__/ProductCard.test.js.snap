// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should render correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginVertical": 15,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <View>
        <Image
          source={
            Object {
              "uri": "https://lp2.hm.com/hmgoepprod?set=source[/83/b3/83b356d257923dc276fc2138e29370e7bf22b7c1.jpg],origin[dam],category[],type[DESCRIPTIVESTILLLIFE],res[m],hmver[2]&call=url[file:/product/style]",
            }
          }
          style={
            Object {
              "borderRadius": 30,
              "height": 150,
              "width": 150,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flex": 1,
          "flexGrow": 1,
          "marginHorizontal": 15,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
              "fontFamily": "Balsamiq Sans",
            },
            Object {
              "marginVertical": 10,
            },
          ]
        }
      >
        Oversized Hooded Jacket
         
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "Balsamiq Sans",
                },
                Object {
                  "color": "#ee7e58",
                  "fontSize": 25,
                  "fontWeight": "bold",
                  "marginEnd": 7,
                },
              ]
            }
          >
            $
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "Balsamiq Sans",
                },
                Object {
                  "fontSize": 17,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            29.99
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontFamily": "Balsamiq Sans",
              },
              Object {
                "flexDirection": "row",
                "justifyContent": "flex-start",
                "marginVertical": 5,
              },
            ]
          }
        >
          Size: 
        </Text>
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "flex-start",
              "marginVertical": 5,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "Balsamiq Sans",
                },
                undefined,
              ]
            }
          >
            Color: 
          </Text>
          <View
            style={
              Object {
                "backgroundColor": undefined,
                "borderColor": "#aaa",
                "borderRadius": 10,
                "borderWidth": 1,
                "height": 20,
                "marginBottom": 10,
                "marginEnd": 10,
                "width": 20,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
